### PR TITLE
MLIBZ-3145: Failing in a batch of multi-insert requests

### PR DESCRIPTION
**Pull request by eugeninnowise**
#### Description
In a batch of multi-insert requests, if a whole batch fails, the response to the whole operation is an error.

#### Changes
The logic of handling exceptions in the multi-insert operation.

#### Tests
TestSaveMultiInsert201CountWithErrorsAsync()

### CODE REVIEW

File: Kinvey.Shared/Store/MultiInsertRequest.cs [hunk 1]
```diff
@@ -362,19 +362,17 @@ private async Task<KinveyMultiInsertResponse<T>> HandleMultiInsertRequestAsync(I
...
+                    throw;
                 }
-                else
+
+                response.Entities = new List<T>();
+                response.Errors = new List<Error>();
+                for (var index = 0; index < entities.Count(); index++)
                 {
-                    throw;
+                    response.Entities.Add(default(T));
```
Comment (heyzooi) [inline Kinvey.Shared/Store/MultiInsertRequest.cs]: @vinaygahlawat correct me if I'm wrong, but I believe a general error (meaning not an error returned by one of the items in the array) should be handled as a regular general error (meaning without wrap the error inside an error with `index` property for example) right?

Comment (vinaygahlawat) [reply Kinvey.Shared/Store/MultiInsertRequest.cs]: @heyzooi It is an interesting question. In the case with a single multi-insert request that fails as a general error this is correct. But what if, in a set of multi-insert requests that are batched, one of the multi-insert operations fail, but the rest succeed. Should we fail the entire operation? Keep in mind that one or several of the batches may have already been processed successfully.

Comment (heyzooi) [reply Kinvey.Shared/Store/MultiInsertRequest.cs]: oh, sorry I didn't notice that we were talking about a multi batch request. So in this case that's correct, we should wrap the error.

File: Kinvey.Shared/Store/MultiInsertRequest.cs [hunk 2]
```diff
@@ -362,19 +362,17 @@ private async Task<KinveyMultiInsertResponse<T>> HandleMultiInsertRequestAsync(I
...
+                    response.Errors.Add(new Error { Code = 0, Errmsg = exception.Message, Index = index });
```
Comment (heyzooi) [inline Kinvey.Shared/Store/MultiInsertRequest.cs]: in this case, the `Index` value should be relative to the batch right? for example let's say the 1st batch works just fine and the 2nd batch fails, in this case the `Index` value should start with `100` right?

Comment (eugeninnowise) [reply Kinvey.Shared/Store/MultiInsertRequest.cs]: @heyzooi Yes, it is correct.
We are resolving all indexes according to an initial batch in code implementation below.
https://github.com/Kinvey/dotnet-sdk/blob/master/Kinvey.Shared/Store/MultiInsertRequest.cs#L309

Comment (heyzooi) [reply Kinvey.Shared/Store/MultiInsertRequest.cs]: oh ok, looks good! Not sure if can be done the same way here, but in Swift we add the `errors` before the `entities` and use the `entities.count` as a offset for the error index, so we don't need to iterate twice just to fix the index. Here's how is done https://github.com/Kinvey/swift-sdk/blob/e4dd8eb1255d6f1e098a40f095d1b81bd416212d/Kinvey/Kinvey/SaveMultiOperation.swift#L293

Comment (eugeninnowise) [reply Kinvey.Shared/Store/MultiInsertRequest.cs]: @heyzooi Thank you for the the advice!
I will look into Swift implementation.

**Pull request merged**